CONTRIBUTING.md

```diff
@@ -1,70 +1,100 @@
-[fork the repo!]: https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/cloning-and-forking-repositories-from-github-desktop
-[Fork the repo]: https://docs.github.com/en/github/getting-started-with-github/fork-a-repo#fork-an-example-repository
+[forking the repository]: https://docs.github.com/en/github/getting-started-with-github/fork-a-repo
 [rustup.rs]: https://rustup.rs/
 
-### Commit Your Update
+#### What You Should Do:
+
+* **API changes**: If your contribution changes the API, please adapt the driver (including module level documentation) and examples accordingly and update the [HIL] (Hardware-in-the-Loop) tests.
+* **Run Related Examples**: After making changes, run any affected examples to ensure they build successfully and perform as expected.
+* **Manual Testing**: For hardware-related changes, manually test your changes on the actual devices when possible. If not, please note it in the corresponding issue, and someone from our team will assist with testing. This is crucial because hardware behavior can sometimes differ from what's simulated or expected.
+* **HIL Tests**: Ensure that any changes to the API or hardware interaction logic are reflected in the HIL tests located in the `hil-test` directory. This helps verify the real-world applicability of your changes.
+
+By taking these extra steps to test your contributions, you help maintain the high quality and reliability of `esp-hal`, ensuring it remains a robust platform for everyone.
 
-Commit the changes once you are happy with them. Don't forget to self-review to speed up the review process.
+[HIL]: https://github.com/esp-rs/esp-hal/tree/main/hil-test
 
-We ask that you ensure all source code files has been properly formatted with `rustfmt`, and that you have linted your changes by running `cargo clippy`. These tools can be installed by running the following commands:
+### Testing Your Contributions
+
+Ensuring the quality and reliability of `esp-hal` is a shared responsibility, and testing plays a critical role in this process. Our GitHub CI automatically checks the buildability of all examples and drivers within the project. However, automated tests can't catch everything, especially when it comes to the nuanced behavior of hardware interactions. So make sure that the example affected by your change works as expected. 
+
+Further steps that can (or should) be taken in testing: 
+
+* Using [xtask], build examples for the specified chip. 
+* Build the documentation if it has been modified using the `run-doc-test` command in [xtask].
+* Run the [HIL] tests locally if changes have been made to them. 
+
+[xtask]: https://github.com/esp-rs/esp-hal/tree/main/xtask
+
+### Commit Your Updates
+
+Commit your changes once you're satisfied. Review your own work to streamline the review process later. Use `rustfmt` and `cargo clippy` to ensure your code adheres to Rust's conventions.
```

Comment:
We should suggest using `lint-packages` and `fmt-packages` `xtask `subcommands

Reply:
`fmt-packages` is already suggested (for some reason I posted a while back that it was `fmt-workspace`, so thanks for the comment, typo corrected) 

`lint-packages` suggestion was also added